### DISCUSSION TITLE
How to manage external dependencies for a publishable library?

### DISCUSSION THREAD

**Discussion by askides** [nrwl/nx#10833] [General]
Hello guys,

I'm using NX for a while, but now I encountered a problem that I've never handled before.

The question is: I have a publishable library which is a CLI application,  in my workspace, and this library requires some external dependencies to be installed in order to work correctly.

So in order to publish the library with the correct dependency, I manually copied the required dependencies from the `global package.json` in my workspace, and I put those into the `specific package.json` of my library. 

I'm struggling to understand if this is the correct way to handle this issue, or if there is a better one. Because, it works, but in case i would update a package version into my global `package.json` i need to manually repeat this replacing process also into the library `package.json`.

Thank you so much.

Renny

**Comment by StevanCurcic** [comment]
i have like 45 publishable libs, all of them with their own unique little dependencies, and its a real hassle to maintain dependencies both in their own pkg.json and the main pkg.json by hand. i'd love to hear the recommended way to handle this.

**Reply by mattfysh** [reply]
This seems like a common problem and nx is positioned to solve it... and yet still there's no word from the team on how to handle this scenario. I've been looking at the nx repo itself to understand how they approach it, and they are also (presumably) copying entries by hand and keeping them in sync manually, e.g.

1. https://github.com/nrwl/nx/blob/85a5c89485f5524ef4119e698872cd203dde26d3/package.json#L160
2. https://github.com/nrwl/nx/blob/85a5c89485f5524ef4119e698872cd203dde26d3/packages/jest/package.json#L43

It's a little curious to see they are having this pain point themselves, but haven't added anything to nx to address it. On top of that, there is also nothing available for version bumping multiple packages in lockstep. Again, having looked into the official nx repo, they are using lerna to version and publish packages -- I'd be interested to hear why these features haven't been ported to nx if the team themselves are finding them useful for managing their own monorepo

**Comment by ydomenjoud** [comment]
Hi !

same thoughts here.

Appreciate some inputs on this.

Thanks

**Comment by gi** [comment]
Try adding your publishable library/package to the root `package.json` as a local path:

```json
"@workspace-name/lib-name": "file:packages/lib-name"
```

That should pull the dependencies of that package into the root `package-lock.json` and `node_modules`.

**Reply by StevanCurcic** [reply]
i'll be sure to, as soon as i can grab some time to experiment. thanks for the input, i really hope it works 👍

**Reply by HalahRaadSalih** [reply]
I've tried your suggestion @gi , but that didn't make a difference. 

I'd appreciate some input/recommendations from the NX team.

**Comment by AdditionAddict** [comment]
I'm curious if `getDependentPackagesForProject` available in "^15.1.0-beta.2" has been created for this purpose?
But yes, this is a huge pain point. I'm trying to update the somewhat dated `nx-firebase` (https://github.com/simondotm/nx-firebase) so I can have easy firebase functions app. This means copying files to `dist`. It'd be nice to have access to the tree. Or is there a way to run a generator inside an executor?

**Comment by KrishnaPG** [comment]
Came across this while looking for a solution for the same problem. What is the recommended way to organizate the dependencies in local packages vs the global package.json
